subtests/docker_cli/login/login.py

```diff
@@ -0,0 +1,333 @@
+                         % (server, auth_file)))
+        else:
+            # Stored credentials
+            cred_hash = creds['auths'][server]['auth']
+            # ...simply a base64-encoded('username:password'), see:
+            #  https://coreos.com/os/docs/latest/registry-authentication.html
+            expected = b64encode('{user}:{passwd}'.format(**self.sub_stuff))
+            self.failif_ne(cred_hash, expected,
+                           'saved credentials in %s' % auth_file)
+
+    def docker_push(self, img_name, tag_name):
+        """
+        Run docker tag, then docker push against our local registry.
+        Tag should always succeed but make no assumptions about push.
+        """
+        push_image = self.config['push_image']
+        dc = DockerCmd(self, 'pull', [push_image])
+        mustpass(dc.execute())
+        self.sub_stuff['my_images'] = [push_image]
+        # FIXME: add push_image to self.containers_needing_cleanup
+
+        self.sub_stuff['img_name'] = img_name
+        self.sub_stuff['tag_name'] = tag_name
+        pushed_name = ('{server}/{img_name}:{tag_name}'
+                       .format(**self.sub_stuff))
+        dc = DockerCmd(self, 'tag', [push_image, pushed_name])
+        mustpass(dc.execute())
+        self.sub_stuff['my_images'] += [pushed_name]
+
+        return DockerCmd(self, 'push', [pushed_name]).execute()
+
+    def cleanup(self):
+        DockerContainers(self).clean_all(self.sub_stuff['my_containers'])
+        if 'my_images' in self.sub_stuff:
+            DockerImages(self).clean_all(self.sub_stuff['my_images'])
+        self._restore_docker_auth_file()
+
+
+class login_ok(login_base):
+
+    """
+    1. Run 'docker login' with correct password
+    2. Verify zero (success) exit status, and 'Login Succeeded' in stdout
+    """
+
+    def run_once(self):
+        super(login_ok, self).run_once()
+        self.sub_stuff["cmdresult"] = self.docker_login()
+
+    def postprocess(self):
+        super(login_ok, self).postprocess()
+        cmdresult = self.sub_stuff['cmdresult']
+        OutputGood(cmdresult)
+        mustpass(cmdresult)
+        self.failif_not_in("Login Succeeded", cmdresult.stdout,
+                           "stdout from docker login")
+        self._check_credentials()
+
+
+class login_fail(login_base):
+
+    """
+    1. Run 'docker login' with incorrect password
+    2. Verify error exit status, and reason in stderr
+    """
+
+    def run_once(self):
+        super(login_fail, self).run_once()
+
+        wrong_passwd = '!' + self.sub_stuff['passwd']
+        self.sub_stuff["cmdresult"] = self.docker_login(wrong_passwd)
+
+    def postprocess(self):
+        super(login_fail, self).postprocess()
+        cmdresult = self.sub_stuff['cmdresult']
+        OutputGood(cmdresult, ignore_error=True)
+        mustfail(cmdresult, 1)
+        self.failif_not_in("no successful auth challenge", cmdresult.stderr,
+                           "stderr from failed docker login")
```

Comment:
Hmmm, I'm not sure.  It's a fairly well-known problem, so I think liklihood of docker breaking vs the test breaking is not in our hair-length's favor.  I'm fine w/o the extra work, it can always be a TODO.